cylc/flow/network/resolvers.py

```diff
@@ -204,20 +223,91 @@ def __init__(self, schd):
         for nat_id in nat_ids:
             oname, wname, _ = nat_id.split(ID_DELIM, 2)
             w_ids.add(f'{oname}{ID_DELIM}{wname}')
-        if self.schd.ws_data_mgr.workflow.id not in w_ids:
-            return []
-        edges = getattr(self.schd.ws_data_mgr.get_entire_workflow(), 'edges')
         return sort_elements(
             [edge
-             for edge in edges
-             if edge.id in nat_ids],
+             for flow in [self.data[w_id] for w_id in w_ids]
+             for edge in [flow['edges'][e_id]
+                          for e_id in nat_ids
+                          if e_id in flow['edges']]],
             args)
 
+    async def get_nodes_edges(self, root_nodes, args):
+        """Return nodes and edges within a specified distance of root nodes."""
+        # Initial root node selection.
+        nodes = root_nodes
+        node_ids = set(n.id for n in root_nodes)
+        edges = []
+        edge_ids = set()
+        # Setup for edgewise search.
+        new_nodes = root_nodes
+        for _ in range(args['distance']):
+            # Gather edges.
+            # Edges should be unique (graph not circular),
+            # but duplicates will be present as node holds all associated.
+            new_edge_ids = set(
+                e_id
+                for n in new_nodes
+                for e_id in n.edges).difference(edge_ids)
+            edge_ids.update(new_edge_ids)
+            w_ids = set()
+            for edge_id in new_edge_ids:
+                o_name, w_name, _ = edge_id.split(ID_DELIM, 2)
+                w_ids.add(f'{o_name}{ID_DELIM}{w_name}')
+            new_edges = [
+                edge
+                for flow in [self.data[w_id] for w_id in w_ids]
+                for edge in [
+                    flow['edges'][e_id]
+                    for e_id in new_edge_ids
+                    if e_id in flow['edges']]
+            ]
+            edges += new_edges
+            # Gather nodes.
+            # One of source or target will be in current set of nodes.
+            new_node_ids = set(
+                [e.source for e in new_edges]
+                + [e.target for e in new_edges]).difference(node_ids)
+            # Stop searching on no new nodes
+            if not new_node_ids:
+                break
+            node_ids.update(new_node_ids)
+            w_ids = set()
+            for node_id in new_node_ids:
+                o_name, w_name, _ = node_id.split(ID_DELIM, 2)
+                w_ids.add(f'{o_name}{ID_DELIM}{w_name}')
+            new_nodes = [
+                node
+                for flow in [self.data[w_id] for w_id in w_ids]
+                for node in [
+                    flow['task_proxies'][n_id]
+                    for n_id in new_node_ids
+                    if n_id in flow['task_proxies']]
+            ]
```

Comment:
This pattern appears in various places?

Reply:
Yes it replaced iterating over a list with a hash table lookup.